File: ios/EpubToMp3/EpubToMp3/Models/SessionRecord.swift
import Foundation

/// Mirrors a single record returned by `GET /api/sessions`.
///
/// The Python backend (see `python_app/src/session_logger.py`) writes records
/// shaped like:
///   { timestamp, book_title, engine, chapters_converted,
///     duration_seconds, outcome, mode, ... }
///
/// We only decode the fields the iOS UI currently needs. Unknown fields are
/// ignored thanks to `Codable` default behaviour.
struct SessionRecord: Codable, Identifiable, Hashable {
    let timestamp: String
    let bookTitle: String
    let engine: String?
    let chaptersConverted: Int?
    let durationSeconds: Double?
    let outcome: String?
    let mode: String?

    /// Stable identifier for SwiftUI lists. The session log doesn't ship an
    /// explicit id, so the timestamp + title combo is used as a surrogate.
    var id: String { "\(timestamp)|\(bookTitle)" }

    enum CodingKeys: String, CodingKey {
        case timestamp
        case bookTitle = "book_title"
        case engine
        case chaptersConverted = "chapters_converted"
        case durationSeconds = "duration_seconds"
        case outcome
        case mode
    }
}

/// Envelope returned by `GET /api/sessions`.
struct SessionsResponse: Codable {
    let sessions: [SessionRecord]
    let count: Int
}

/// One line of an SSE stream payload coming from `/api/jobs/{id}/stream`.
/// The backend emits a JSON `JobSnapshot` per `data:` line; we surface the
/// raw text here for the slice and let later versions decode the full snapshot.
struct JobEvent: Identifiable, Hashable {
    let id = UUID()
    let receivedAt: Date
    let rawPayload: String
}

File: flutter_app/lib/models/session_record.dart
// ignore_for_file: invalid_annotation_target
import 'package:freezed_annotation/freezed_annotation.dart';

part 'session_record.freezed.dart';
part 'session_record.g.dart';

/// `GET /api/sessions` — see `python_app/src/session_logger.py`.
/// Wire format is snake_case here (legacy log entries).
@freezed
class SessionRecord with _$SessionRecord {
  const SessionRecord._();
  const factory SessionRecord({
    required String timestamp,
    @JsonKey(name: 'book_title') required String bookTitle,
    String? engine,
    @JsonKey(name: 'chapters_converted') int? chaptersConverted,
    @JsonKey(name: 'duration_seconds') double? durationSeconds,
    String? outcome,
    String? mode,
  }) = _SessionRecord;

  factory SessionRecord.fromJson(Map<String, dynamic> json) =>
      _$SessionRecordFromJson(json);

  String get id => '$timestamp|$bookTitle';
}

@freezed
class SessionsResponse with _$SessionsResponse {
  const factory SessionsResponse({
    required List<SessionRecord> sessions,
    required int count,
  }) = _SessionsResponse;

  factory SessionsResponse.fromJson(Map<String, dynamic> json) =>
      _$SessionsResponseFromJson(json);
}
